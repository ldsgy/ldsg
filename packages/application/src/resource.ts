import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
