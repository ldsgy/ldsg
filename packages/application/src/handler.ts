import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
