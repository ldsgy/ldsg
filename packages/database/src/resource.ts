import { Resource, ResourceConstructorParams } from "@ldsg/resource";
import mongoose from "mongoose";
import { DatabaseSpecificResourceSettings } from "./types";

export class DatabaseResource extends Resource<DatabaseSpecificResourceSettings> {
  constructor(
    params: ResourceConstructorParams<DatabaseSpecificResourceSettings>
  ) {
    super(params);

    const { connect } = this;

    connect();
  }

  connect = () => {
    const {
      settings: { uri, connectOptions },
    } = this;

    return mongoose.connect(uri, connectOptions);
  };

  disconnect = () => {
    return mongoose.disconnect();
  };
}
