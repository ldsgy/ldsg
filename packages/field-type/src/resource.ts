import { Resource } from "@ldsg/resource";
import { FieldTypeSpecificResourceSettings } from "./types";

export class FieldTypeResource extends Resource<FieldTypeSpecificResourceSettings> {}
