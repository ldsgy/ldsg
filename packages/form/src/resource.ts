import { ObjectInfo, ObjectResource } from "@ldsg/object";
import { Resource } from "@ldsg/resource";
import { FormSpecificResourceSettings } from "./types";

interface GetFormInfoParams {
  /**
   * Platform
   * Such as mongoose\formily.
   */
  platform: string;
}

export interface FormInfo {
  /**
   * Form Name
   */
  name: string;

  /**
   * Input Object Info
   */
  inputObjectInfo: ObjectInfo;

  /**
   * Output Object Info
   */
  outputObjectInfo: ObjectInfo;
}

type GetFormInfo = (params: GetFormInfoParams) => FormInfo;

export class FormResource extends Resource<FormSpecificResourceSettings> {
  getFormInfo: GetFormInfo = (params) => {
    const { platform } = params;

    const {
      settings: { name },
      getResourcesFromSettings,
    } = this;

    const { inputObjectResource, outputObjectResource, workflowResource } =
      getResourcesFromSettings();

    const inputObjectInfo = (
      inputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const outputObjectInfo = (
      outputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const res: FormInfo = {
      name,
      inputObjectInfo,
      outputObjectInfo,
    };

    return res;
  };
}
