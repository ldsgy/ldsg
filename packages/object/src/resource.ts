import { PlatformsParams } from "@ldsg/field-type";
import { Resource } from "@ldsg/resource";
import { GeneralResourceSettings } from "@ldsg/types";
import _ from "lodash";
import {
  FieldInfo,
  GetFieldInfo,
  ObjectSpecificResourceSettings,
} from "./types";

export interface ExtendedObjectInfo {
  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetExtendedObjectInfoRes {
  /**
   * Extended Object Info
   */
  extendedObjectInfo: ExtendedObjectInfo;
}

type GetExtendedObjectInfo = (
  params?: PlatformsParams
) => GetExtendedObjectInfoRes;

export interface ObjectInfo extends GeneralResourceSettings {
  /**
   * Object Name
   */
  name: string;

  /**
   * Field Info List
   */
  fieldInfoList: FieldInfo[];
}

export interface GetObjectInfoRes {
  /**
   * Object Info
   */
  objectInfo: ObjectInfo;
}

type GetObjectInfo = (params?: PlatformsParams) => GetObjectInfoRes;

export class ObjectResource extends Resource<ObjectSpecificResourceSettings> {
  getExtendedObjectInfo: GetExtendedObjectInfo = (params) => {
    const {
      settings: { extendedResourceIds },
      getFilteredResource,
    } = this;

    const fieldInfoListGroup = _.map(
      extendedResourceIds,
      (extendedResourceId) => {
        const { resource } = getFilteredResource<ObjectResource>({
          id: extendedResourceId,
        });

        const getObjectInfoRes = resource?.getObjectInfo(params);

        const res = getObjectInfoRes?.objectInfo.fieldInfoList || [];

        return res;
      }
    );

    const fieldInfoList = _.flatten(fieldInfoListGroup);

    const extendedObjectInfo: ExtendedObjectInfo = {
      fieldInfoList,
    };

    const res = {
      extendedObjectInfo,
    };

    return res;
  };

  getObjectInfo: GetObjectInfo = (params) => {
    const {
      id,
      settings: { title, description, name },
      getExtendedObjectInfo,
      getFilteredResources,
    } = this;

    const { extendedObjectInfo } = getExtendedObjectInfo(params);

    const { fieldInfoList: extendedFieldInfoList } = extendedObjectInfo;

    const { resources } = getFilteredResources<{ getFieldInfo?: GetFieldInfo }>(
      {
        parentId: id,
      }
    );

    const mapRes = resources.map((resource) => {
      const fieldInfo = resource.getFieldInfo?.(params);

      return fieldInfo;
    });

    const filterRes = _.filter(mapRes, (value) => !_.isUndefined(value));

    const currentFieldInfoList = _.map(filterRes, "fieldInfo");

    const fieldInfoList = extendedFieldInfoList.concat(currentFieldInfoList);

    const objectInfo: ObjectInfo = {
      title,
      description,
      name,
      fieldInfoList,
    };

    const res = {
      objectInfo,
    };

    return res;
  };
}
