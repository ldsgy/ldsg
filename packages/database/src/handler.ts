import { Handler } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { DatabaseResource } from "./resource";
import { DatabaseSpecificResourceSettings } from "./types";

export const handler: Handler<
  [InstantiateResourceParams<DatabaseSpecificResourceSettings>],
  InstantiateResourceRes<DatabaseResource>
> = instantiateResource;
