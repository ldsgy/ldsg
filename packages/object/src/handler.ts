import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { ObjectResource } from "./resource";
import { ObjectSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<ObjectSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<ObjectResource>
> = instantiateResource;
