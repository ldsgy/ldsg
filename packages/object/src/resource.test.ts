import { FieldTypeResource } from "@ldsg/field-type";
import { Handler, HandlerResource } from "@ldsg/handler";
import { ObjectFieldResource } from "@ldsg/object-field";
import { ObjectResource } from "./resource";

const handler: Handler<
  [
    {
      /**
       * Field Properties
       */
      fieldProperties?: any;
      /**
       * Platform
       * Such as mongoose\formily.
       */
      platform: string;
    }
  ],
  {}
> = (params) => {
  const { fieldProperties, platform } = params;

  const { max } = fieldProperties;

  let res;

  switch (platform) {
    case "mongoose": {
      res = {
        type: "String",
        ...(max
          ? {
              maxLength: max,
            }
          : {}),
      };

      break;
    }

    default: {
      res = {};

      break;
    }
  }

  return res;
};

test("object", () => {
  new HandlerResource({
    id: "test-handler",
    kind: "HANDLER",
    parentId: "root",
    settings: {
      title: "测试处理程序",
      description: "",
      code: "",
      dependencies: [],
    },
    handler,
  });

  new FieldTypeResource({
    id: "test-field-type",
    kind: "FIELD_TYPE",
    parentId: "root",
    settings: {
      title: "文本",
      description: "可用来存储各种文本",
      fieldPropertiesSchema: {
        type: "object",
        properties: {
          max: {
            type: "integer",
            title: "配置最长字符",
            description: "长度不可以超过此值",
          },
        },
      },
      handlerResourceId: "test-handler",
    },
  });

  new ObjectFieldResource({
    id: "test-object-field-1",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段1",
      description: "",
      name: "test1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-field-2",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段2",
      description: "",
      name: "test2",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-field-3",
    kind: "OBJECT_FIELD",
    parentId: "test-object",
    settings: {
      title: "测试字段3",
      description: "",
      name: "test",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  const objectResource = new ObjectResource({
    id: "test-object",
    kind: "OBJECT",
    parentId: "root",
    settings: {
      title: "测试对象",
      description: "",
      name: "test-object",
    },
  });

  const getObjectInfoRes = objectResource.getObjectInfo({
    platform: "mongoose",
  });

  expect(getObjectInfoRes).toMatchSnapshot();
});
