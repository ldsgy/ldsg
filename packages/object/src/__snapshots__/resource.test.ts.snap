// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object 1`] = `
{
  "fieldInfoList": [
    {
      "name": "test1",
      "typeInfo": {
        "maxLength": "10",
        "type": "String",
      },
    },
    {
      "name": "test2",
      "typeInfo": {
        "maxLength": "10",
        "type": "String",
      },
    },
    {
      "name": "test",
      "typeInfo": {
        "maxLength": "10",
        "type": "String",
      },
    },
  ],
  "name": "test-object",
}
`;
